clarify runahead limit (#660)
cylc/cylc-doc: src/user-guide/writing-workflows/scheduling.rst

Runahead Limiting
^^^^^^^^^^^^^^^^^

The runahead limit prevents a workflow from getting too far ahead of the oldest
active cycle point.

A cycle is considered to be active if it contains any :term:`active` or

> :term:`incomplete` tasks.

Reviewer: We need to get our story straight on exactly what "active cycle points" (aka the "active window" of the workflow) contain(s):
- literally active tasks (submitted, running)
- incomplete tasks (finished without completing required outputs)
- waiting tasks that are ready to run as far as the task graph is concerned, but are still waiting on
  - clock triggers
  - xtriggers
  - queue release
  - or hold

This isn't the right place to say all of that, but we should put it somewhere and link all references to "active cycle points" to it.

Reviewer 2: What about tasks with partially satisfied dependencies?
I assume an active cycle is one with a task in the n=0 window?

Reviewer: Uh, yes! Especially after the "Cylc set" PR (which ditches the hidden pool)

Author: Ok, it's sounding like we need a new glossary entry to describe the meaning of "active cycle".

The definition should include:
* Active tasks.
* Incomplete tasks.
* Waiting tasks that have one or more of their dependencies satisfied.
* Waiting parentless tasks.

Good?

Reviewer 2: I think we should define an active cycle as one which contains a task in the active ("n=0") window.
This term is already defined in the glossary but is wrong - I think we should update it to include:
* Active tasks
* Incomplete tasks
* Active-waiting tasks (we might need to tweak the existing definition in the glossary for these)
* Waiting tasks with some, but not all, of their pre-requisites met

> Waiting parentless tasks

I don't understand the relevance of parentless here.

Reviewer: Plus (as I said above):
- waiting tasks that are ready to run as far as the task graph is concerned, but are still waiting on: queues, hold, clock and xtriggers

> >    Waiting parentless tasks
> 
> I don't understand the relevance of parentless here.

Parentless tasks get auto-spawned to the runahead limit (although, after @dwsutherland 's extant PR clock-triggered ones won't). That is technically covered by "ready to run as far at the task graph is concerned" but we should document it anyway - might not be obvious to users.

The runahead limit can be configured for a workflow by

> :cylc:conf:`[scheduling]runahead limit`. It defines the span of cycles which
> are permitted to run as measured from the oldest active cycle point. The
> runahead limit can be defined in two ways, as an integer or datetime span.

Reviewer: ```suggestion
:cylc:conf:`[scheduling]runahead limit`. It can be defined in two ways, as an integer or datetime interval.
```
That repeats the information just above.

.. rubric:: Integer Format

The integer format ``Pn`` is an interval that spans ``n+1`` consecutive cycle
points, regardless of how the cycle point values increment from one point to

> the next (i.e. regardless of the cycling interval).

Reviewer 2: I'm not really sure what "regardless of the cycling interval" means here.
Maybe just say:
"The integer format ``Pn`` is an interval that spans ``n+1`` consecutive cycle points as defined by the graph"

Reviewer: > I'm not really sure what "regardless of the cycling interval" means here

Th intention was, I think, to emphasize that a Pn limit determines the number of cycle points regardless of how big or small the interval between cycles is. (As compared to a duration limit, where the number of cycles will depend on the interval).

> For example, the default runahead limit of ``P4`` allows 5 active cycle points

Reviewer 2: I think "up to 5" would be better

(the oldest active cycle plus the 4 next cycles).

In the following example, tasks ``1/foo`` through ``4/foo`` will submit to

> run immediately at start-up, but ``5/foo`` will be held back until ``1/foo``

Reviewer 2: No! The example uses P2 at the moment. Either correct these or change the example to use P1?
